devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java

```diff
@@ -0,0 +1,299 @@
+            JavaPluginConvention javaConvention = convention.findPlugin(JavaPluginConvention.class);
+            if (javaConvention == null) {
+                continue;
+            }
+            modules.add(getWorkspaceModule(subproject));
+        }
+        return modules;
+    }
+
+    private WorkspaceModule getWorkspaceModule(Project project) {
+        ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
+                project.getVersion().toString());
+        final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+
+        return new WorkspaceModuleImpl(appArtifactCoords, project.getProjectDir().getAbsoluteFile(),
+                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), convert(mainSourceSet));
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getEnforcedPlatforms(Project project) {
+        final Set<org.gradle.api.artifacts.Dependency> directExtension = new HashSet<>();
+        // collect enforced platforms
+        final Configuration impl = project.getConfigurations()
+                .getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME);
+        for (org.gradle.api.artifacts.Dependency d : impl.getAllDependencies()) {
+            if (!(d instanceof ModuleDependency)) {
+                continue;
+            }
+            final ModuleDependency module = (ModuleDependency) d;
+            final Category category = module.getAttributes().getAttribute(Category.CATEGORY_ATTRIBUTE);
+            if (category != null && Category.ENFORCED_PLATFORM.equals(category.getName())) {
+                directExtension.add(d);
+            }
+        }
+        return directExtension;
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getDirectExtensionDependencies(Set<ResolvedDependency> dependencies,
+            Map<ArtifactCoords, Dependency> appDependencies, boolean firstLevel, Set<ArtifactCoords> visited) {
+        Set<org.gradle.api.artifacts.Dependency> extensions = new HashSet<>();
+        for (ResolvedDependency d : dependencies) {
+            ArtifactCoords key = new ArtifactCoordsImpl(d.getModuleGroup(), d.getModuleName(), "");
+            if (!visited.add(key)) {
+                continue;
+            }
+
+            Dependency appDep = appDependencies.get(key);
+            if (appDep == null) {
+                continue;
+            }
+            final org.gradle.api.artifacts.Dependency deploymentArtifact = getDeploymentArtifact(appDep);
+
+            boolean addChildExtension = firstLevel;
+            if (deploymentArtifact != null && addChildExtension) {
+                extensions.add(deploymentArtifact);
+                addChildExtension = false;
+            }
+
+            final Set<ResolvedDependency> resolvedChildren = d.getChildren();
+            if (!resolvedChildren.isEmpty()) {
```

Comment:
I had `addChildExtension` and walked the whole tree because we also used to collect all the extension properties. Given that you are doing it elsewhere, I think you could have changed this if to `addChildExtension && !resolvedChildren.isEmpty()`. I don't think you need `firstLevel` arg in your impl.

Reply:
I.e. you could have removed `addChildExtension` too. We are simply collecting the first met extension dependencies but not its children.